### ISSUE TITLE
race conditions with setPermissions

### ISSUE THREAD

**Issue by setharnold** [moby/moby#43725]
https://github.com/moby/moby/blob/74286cba8c2c14d5360caec44d2da74858e9c707/pkg/idtools/idtools_unix.go#L221

```
func setPermissions(p string, mode os.FileMode, uid, gid int, stat *system.StatT) error {
	if stat == nil {
		var err error
		stat, err = system.Stat(p)
		if err != nil {
			return err
		}
	}
	if os.FileMode(stat.Mode()).Perm() != mode.Perm() {
		if err := os.Chmod(p, mode.Perm()); err != nil {
			return err
		}
	}
	if stat.UID() == uint32(uid) && stat.GID() == uint32(gid) {
		return nil
	}
	return os.Chown(p, uid, gid)
}
```

Hello, I'm concerned about this `setPermissions()` function; by taking a path to files or directories, rather than a file descriptor, it is possible that an untrusted process on the system can race these calls and cause this process to change ownership or permissions on files or directories of their choosing.

Best is to create the files or directories with the correct ownership and permission in the first place; it is extremely difficult for root-privileged processes to operate safely in directories that can be written by untrusted users.

I'd like to suggest that the `mkdirAs()` function be modified to run with the privileges of the desired user, which can reduce the chances for trouble and obviate the need for some of the operations.

Of course, if this routine is never executed on computers with untrusted users, it's not a big deal.

Thanks

**Comment by setharnold** [comment]
Thanks @corhere for the detailed look! I'm very sorry I missed the security policy; on my view of https://github.com/moby/moby the `Security` tab is hidden behind the dreaded hamburger menu. This is news to me. (I've become lazy, too, apparently.)

Thanks again

**Comment by thaJeztah** [comment]
Thank you both! Reading the discussion above, can we conclude that no changes to the code are needed _for the intended purpose of the code_ ?

Are there perhaps parts of the code that need / could use some additional documentation? It can be useful to be explicit about the intended purpose of a function, especially because Go's ecosystem makes it easy to consume code from other projects (such as this repository), but possibly use it for purposes that it was not designed for (see also https://github.com/moby/moby/pull/43477, where I tried to do so).

Of course we could move such code to an `internal/` package, but I must admit that I'm also not _always_ a fan of that; in some cases it's perfectly fine to use the code _as long as consumers know the limitations / purpose_ (allowing them to decide whether or not those fit their intended use).

**Comment by corhere** [comment]
Yes, I agree with your conclusions: aside from some additional documentation, no code changes are needed.

**Comment by setharnold** [comment]
@thaJeztah that sounds like a good outcome to me; it's unrealistic to ask for every piece of code to be made resilient against every possible contingency, and sometimes a big warning: "not for use when untrusted processes can manipulate the filesystem" is about the best you can do. (And sadly, given the system calls available to us, this is a problem common to basically everything in the ecosystem.)

Thanks
